NODE_METADATA_DESIGN.md

```diff
@@ -0,0 +1,49 @@
+## Design proposal for additional node metadata - [JENKINS-26522](https://issues.jenkins-ci.org/browse/JENKINS-26522)
+
+### Problem statement
+
+Before this change, `FlowNode`s only can have a limited amount of metadata. For example, status is simply determined by
+whether there is an `ErrorAction` on the `FlowNode` - if so, it's considered failed, otherwise, it's considered passed.
+There's no way to record or report on more granular status, such as `UNSTABLE` or `ABORTED` for an individual `FlowNode`
+(or block/stage). There also is no way to allow plugins to contribute additional metadata to `FlowNode`s, such as test
+results, so there's no way to report on which particular step or stage contributed which test results. Combined, this
+limits the reporting and visualization capabilities available to Blue Ocean and the Pipeline Stage View significantly.
+
+### Summary of Design
+
+To address this, we will use a new `PersistentAction` similar to `TagsAction` to record that needed additional metadata.
+This new action, named something like `PropagatingTagsAction` to signify that it plays a similar role to `TagsAction`, 
+with the noteworthy difference that its tags are propagated to enclosing blocks. Status in particular will be 
+special-cased in `PropagatingTagsAction` to enforce the rule that status can only get worse, not better. 
+
+`PropagatingTagsAction` will support multiple `String` values for a single tag, to support cases such as aggregating 
```

Comment:
So seeing the APIs laid out here formally (thank you!) actually makes me realize there's a couple questions we need to answer collectively:

* How much logic do we want to associate with a tag?  I've been assuming it's trivial, but if not, then we do want a NodeTagExtension that provides APIs which interact with the TagAction/ScopedTagAction
* How much are we storing custom data here?  My gut is that this needs to be pretty light in general, otherwise we're going to be accumulating a lot of crap in the FlowGraph that might better be attached to the Run.